File: from-1701-to-1800/1717-maximum-score-from-removing-substrings/solution.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def remove_pairs(pair: str, score: int) -> int:
            nonlocal s
            result = 0
            stack = []
            for c in s:
                if c == pair[1] and stack and stack[-1] == pair[0]:
                    stack.pop()
                    result += score
                else:
                    stack.append(c)

            s = "".join(stack)
            return result

        result = 0
        pair = "ab" if x > y else "ba"
        result += remove_pairs(pair, max(x, y))
        result += remove_pairs(pair[::-1], min(x, y))

        return result

File: from-1701-to-1800/1717-maximum-score-from-removing-substrings/solution.go
func maximumGain(s string, x int, y int) int {
	removePairs := func(pair string, score int) int {
		result := 0
		stack := []byte{}
		for i := 0; i < len(s); i ++ {
            if s[i] == pair[1] && len(stack) > 0 && stack[len(stack) - 1] == pair[0] {
                stack = stack[:len(stack) - 1]
                result += score
            } else {
                stack = append(stack, s[i])
            }
		}
        s = string(stack)
        return result
	}

    result := 0
    p1, p2 := "ab", "ba"
    s1, s2 := x, y
    if s2 > s1 {
        p1, p2 = p2, p1
        s1, s2 = s2, s1
    }
    result += removePairs(p1, s1)
    result += removePairs(p2, s2)
    return result
}